gc.c

```diff
@@ -8209,6 +8010,166 @@ gc_marks_step(rb_objspace_t *objspace, size_t slots)
 }
 #endif
 
+static bool
+gc_compact_heap_cursors_met_p(rb_heap_t *heap)
+{
+    if (heap->sweeping_page && heap->compact_cursor) {
+        return heap->sweeping_page == heap->compact_cursor;
+    } else {
+        return false;
+    }
+}
+
+static void
+gc_compact_sweep_unswept_page(rb_objspace_t *objspace, rb_heap_t *heap, struct heap_page *page)
```

Comment:
I feel like we can combine this function with `try_move`, and do something like this:

1. If we don't have a free page:
    a. Return false if the cursors have met
    b. Sweep the page at the `sweeping_page`
    c. Move `sweeping_page` cursor forward
    d. Goto step a if page does not have any free slots
2. Move `src` to slot in free page

I think that it should be an invariant that `sweeping_page` is pointing to a list of unswept pages rather than having to check if the page has been swept or not.